### PR TITLE
Socint 154 new case event

### PR DESCRIPTION
**Pull request by SteveScorfield**
# Motivation and Context
The change was required for adding the new case creation code.

# What has changed
Code relating to new case creation has been added along with alterations to existing code to fit in with changes made to common.

Yes

Tests relating to the new case creation have been added along with updating existing tests

# How to test?

Ran them locally with the unit tests, moved to acceptance tests, test locally using an example JSON and have deployed to dev.

# Links

https://collaborate2.ons.gov.uk/jira/browse/SOCINT-154

https://collaborate2.ons.gov.uk/confluence/display/SDC/New+Case+Event

### CODE REVIEW

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImpl.java [hunk 1]
```diff
@@ -100,6 +104,51 @@ public void fulfilmentRequestByPost(PostalFulfilmentRequestDTO requestBodyDTO)
...
+    sampleSensitive.setChildDob(caseRegistrationDTO.getChildDob());
+    sampleSensitive.setAdditionalInfo(caseRegistrationDTO.getAdditionalInfo());
+    sampleSensitive.setChildMobileNumber(caseRegistrationDTO.getChildMobileNumber());
+    sampleSensitive.setChildEmailAddress(caseRegistrationDTO.getChildEmailAddress());
+    sampleSensitive.setParentMobileNumber(caseRegistrationDTO.getParentMobileNumber());
+    sampleSensitive.setParentEmailAddress(caseRegistrationDTO.getParentEmailAddress());
+
+    NewCasePayloadContent newCasePayloadContent =
+        new NewCasePayloadContent(caseId, collectionExerciseId, sample, sampleSensitive);
+    ;
```
Comment (rpcatling) [inline src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImpl.java:147]: random extra semicolon

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImpl.java [hunk 2]
```diff
@@ -100,6 +104,51 @@ public void fulfilmentRequestByPost(PostalFulfilmentRequestDTO requestBodyDTO)
...
+    eventPublisher.sendEvent(TopicType.NEW_CASE, Source.RESPONDENT_HOME, Channel.RH, payload);
+  }
+
+  private NewCasePayloadContent createNewCaseRequestPayload(NewCaseDTO caseRegistrationDTO)
+      throws CTPException {
+
+    final UUID caseId = UUID.randomUUID();
+    final UUID collectionExerciseId = caseRegistrationDTO.getCollectionExerciseId();
+
+    NewCaseSample sample = new NewCaseSample();
```
Comment (rpcatling) [inline src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImpl.java:126]: The normal pattern for these mappings is to use the bean mapper `RHSvcBeanMapper`.

File: src/test/java/uk/gov/ons/ctp/integration/rhsvc/message/impl/CaseEventReceiverImplIT_Test.java [hunk 3]
```diff
@@ -73,9 +78,13 @@ public void caseCaseReceivedWithoutMillisTest() throws Exception {
...
     CaseEvent caseEvent = FixtureHelper.loadPackageFixtures(CaseEvent[].class).get(0);
 
+    Header header = new Header();
     SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.ENGLISH);
     sdf.setTimeZone(TimeZone.getTimeZone("GMT"));
-    caseEvent.getEvent().setDateTime(sdf.parse("2011-08-12T20:17:46Z"));
+    header.setDateTime(sdf.parse("2011-08-12T20:17:46Z"));
+    header.setMessageId(UUID.fromString("c45de4dc-3c3b-11e9-b210-d663bd873d93"));
+    caseEvent.setHeader(header);
+    //    caseEvent.getHeader().setDateTime(sdf.parse("2011-08-12T20:17:46Z"));
```
Comment (rpcatling) [inline src/test/java/uk/gov/ons/ctp/integration/rhsvc/message/impl/CaseEventReceiverImplIT_Test.java:87]: not sure why this is commented out ?

File: src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImplTest.java [hunk 4]
```diff
@@ -175,4 +187,61 @@ public void getCaseByUPRNNotFound() throws Exception {
...
+    assertEquals(newCaseDTO.getChildDob(), eventPayload.getSampleSensitive().getChildDob());
+    assertEquals(
+        newCaseDTO.getParentMobileNumber(),
+        eventPayload.getSampleSensitive().getParentMobileNumber());
+    assertEquals(
+        newCaseDTO.getParentEmailAddress(),
+        eventPayload.getSampleSensitive().getParentEmailAddress());
+  }
+
+  public NewCaseDTO getNewCaseDto() {
```
Comment (rpcatling) [inline src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImplTest.java:230]: This is not our normal fixture method. I'm not a fan, but our standard pattern is to create a JSON file and covert it to an Object with `FixtureHelper` eg, like the collectionCase above in this file.

File: src/test/resources/uk/gov/ons/ctp/integration/rhsvc/endpoint/CaseEndpointUnitTest.NewCaseEventDTO.json [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  "childFirstName": "Jo",
+  "childMiddleName": "Rose",
+  "childLastName": "Bloggs",
+  "childDob": "2001-12-31",
+  "additionalInfo": "Class 2A",
+  "childMobileNumber": "447123456777",
+  "childEmailAddress": "jo.rose.bloggs@domain.com",
+  "parentMobileNumber": "447123456999",
+  "parentEmailAddress": "fred.bloggs@domain.com"
+}
```
Comment (rpcatling) [inline src/test/resources/uk/gov/ons/ctp/integration/rhsvc/endpoint/CaseEndpointUnitTest.NewCaseEventDTO.json:19]: please add a new line to remove the little red warning in github.

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/representation/NewCaseDTO.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+  private int surveyId;
+  @NotNull private UUID collectionExerciseId;
+
+  @NotBlank private String schoolId;
+  @NotBlank private String schoolName;
+
+  private boolean consentGivenTest;
+  private boolean consentGivenSurvey;
+
+  @NotBlank private String firstName;
```
Comment (rpcatling) [inline src/main/java/uk/gov/ons/ctp/integration/rhsvc/representation/NewCaseDTO.java:29]: It's not clear to me if more of these should be marked with 
```
@LoggingScope(scope = Scope.MASK)
```

Which ones are to be masked ? if they are all the ones in the "sampleSensitive" list then things like `firstName` and `surname` and several others need to be included

File: src/test/java/uk/gov/ons/ctp/integration/rhsvc/endpoint/CaseEndpointUnitTest.java [hunk 7]
```diff
@@ -312,6 +313,35 @@ public void shouldRejectFulfilBySmsWithIncorrectRequestBody() throws Exception {
     verifyRejectedSmsFulfilmentRequest(json, url);
   }
 
+  /** Test should send out a new case event (register) */
```
Comment (rpcatling) [inline src/test/java/uk/gov/ons/ctp/integration/rhsvc/endpoint/CaseEndpointUnitTest.java:316]: In order to test all the validation rules using `javax.validation.constraints` annotations in the `NewCaseDTO` we need quite a few more unit tests here, eg `firstName` cannot be blank etc etc

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImpl.java (hunk 2 above)
Comment (SteveScorfield) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImpl.java]: This has been changed to use the RHSvcBeanMapper

**Review by rpcatling — APPROVED** [review]

**Pull request merged**